### ISSUE TITLE
Why are there differences between models trained through different training methods?

### ISSUE THREAD

**Issue by AllenSun1024** [microsoft/LightGBM#5803]
## Description
Given **the same hyperparameters** and **the same training data**, there are some differences between models trained through different training methods.
I've tried the following two methods to train with same hyperparameters and same training data: 

1. train a LGBMRegressor model in CLI
2. train a LGBMRegressor model with lgb.sklearn API in Python

## Reproducible example

- about training data: 
    - label column is the last column
    - CLI version uses **csv format** while Python version uses **parquet format**
- about hyperparameters:
    - CLI version: 
      - label_column = 3210
      - data = train_data.csv
      - valid = test_data.csv
      - colsample_bytree = 0.1
      - n_estimators = 5000
      - n_jobs = 20
    - Python version:
      - colsample_bytree = 0.1
      - n_estimators = 5000
      - n_jobs = 20

1. To reproduce CLI version:
```shell
./lightgbm config=CLI_train.conf
```

2. To reproduce Python API version:
```python
data = pandas.read_parquet("train_data.par")
model = lightgbm.sklearn.LGBMRegressor(**hyper_params_dict)
model.fit(data.iloc[:, :-1].values, data.iloc[:, -1].values)
model.booster_.save_model("py_model.txt")
```

## Environment info

LightGBM version: lightgbm 3.3.5.99

Command(s) you used to install LightGBM

```shell
git clone --recursive https://github.com/microsoft/LightGBM
cd LightGBM
mkdir build
cd build
cmake ..
make -j$(nproc)
cd ..
cd python-package
python setup.py install --precompile
```

## Additional Comments
**The differences between models** trained by CLI version and Python version  are shown below:

1. feature_infos
2. tree_sizes
3. Trees[0, 1, 2, ..., n_estimators - 1]
4. feature_importances
5. label_index: 3210(CLI) / 0(Python) **It's quite weird that the label index of Python version should be 0 since we've set it equivalently through data.iloc[:, -1].values in python. From my point of view, the model parameter `label_index` should be meaningless if the model is trained through Python version.**

The random seeds(i.e., bagging_seed, feature_fraction_seed, ...) are all set default values.

After preliminary exploration, I think **the following reasons may lead to the model difference**: 

1. CLI tools can control **the parallel processing mode** of LightGBM by setting parameters such as OpenMP. However, the Scikit-learn style API uses Scikit-learn's parallel processing facilities to achieve parallelization.
2. Lightgbm.sklearn.LGBMRegressor provides two **built-in methods for cross-validation and hyperparameter tuning**: GridSearchCV and RandomizedSearchCV while the CLI version of LightGBM does not provide built-in methods for cross-validation and hyperparameter tuning.

_**Am I right?**_

Model info is listed here:
tree
version=v4
num_class=1
num_tree_per_iteration=1
label_index=[3210(CLI) | 0(Python)]
max_feature_idx=3209
objective=regression
feature_names=...
feature_infos=...
tree_sizes=...
...

**Comment by AllenSun1024** [comment]
I'm sorry for committing this issue. 

**Models trained through CLI and Python are same**.

The reason is listed below:

Training data used in Python is parquet format, which contains header line and this line won't be considered as a sample data.

`We converted parquet data into csv straightly without deleting the header line and it lead to a bug.`

The header line should not exist when we try to train model through CLI.
